Cloud Platform (#64)
OctopusDeploy/Architecture: CloudPlatform/BackgrundProcessesAndTasks.md

# Summary #

Background processes are a custom implementation of [infrastructure as code (IaC)](https://en.wikipedia.org/wiki/Infrastructure_as_code) approach to managing infrastructure. 

You can think about background processes as classes in C#. Background tasks are instances of those classes. 

If you are familiar with NServiceBus then you can think about Background Processes as [NServiceBus Sagas](https://docs.particular.net/nservicebus/sagas/) 

# Details #

## Background Processes ##

Each background process consists of a set of steps. Each step must be [idempotent](https://en.wikipedia.org/wiki/Idempotence).

Each background process has state that is shared by all its steps. 

## Background Tasks ##

Background tasks are the runtime representation of background processes. Each task is broken down into steps that are executed sequentially in the order they are defined in the process.

> The task runner (should we call it step runner?) runs up to 35 steps from different tasks in parallel, in the order in which they where scheduled. This means that there might be hundreds of running tasks in the system at any point of time. You can think about steps as units of work.

Reviewer: "(should we call it step runner?)" -> I'd delete this

Author: 👍

> Every step is automatically retried (up to 3 times) when it fails which is why all steps need to be idempotent.

Reviewer 2: Also mention that if a step is interrupted due to an app restart, it will be retried?

Author: 👍

> The step that will be executed next in a given task can be manually overridden. This is a very powerful feature but it needs to be used carefully to make sure the task's state won't be corrupted.

Reviewer: Would be good to expand here as we're asking people to be cautious. 
"This feature is useful for resuming a task from a specific step. If you have any doubt about the likelihood or impact of corrupting the task state, don't use this manual override feature."

Author: 👍

> The step that will be executed next in a given task can be manually overridden. This is a very powerful feature but it needs to be used carefully to make sure the task's state won't be corrupted.

Reviewer: ```suggestion
The step that will be executed next in a given task can be manually overridden. This feature is useful for resuming a task from a specific step. This is a very powerful feature but it needs to be used carefully to make sure the task's state won't be corrupted. If you have any doubt about the likelihood or impact of corrupting the task state, don't use this manual override feature.
```

## FAQ ##

1. Why didn't you use Terraform?

> Initially, reef's resources were managed using TF scripts. Unfortunately, TF wasn't flexible enough for us. [Pulumi](https://www.pulumi.com/) is a tool we should have a look if we wanted to replace our custom solution with a 3rd party tool.

Reviewer: "have a look" -> change to "consider"

Author: 👍

> Initially, reef's resources were managed using TF scripts. Unfortunately, TF wasn't flexible enough for us. [Pulumi](https://www.pulumi.com/) is a tool we should have a look if we wanted to replace our custom solution with a 3rd party tool.

Reviewer: ```suggestion
Initially, reef's resources were managed using TF scripts. Unfortunately, TF wasn't flexible enough for us. [Pulumi](https://www.pulumi.com/) is a tool we should consider if we want to replace our custom solution with a 3rd party tool in the future.
```
